src/vdaf/mastic.rs

```diff
@@ -379,32 +382,477 @@ where
+            (None, Some(verifier_len)) => Ok(Some(
+                SzkQueryShare::<F, SEED_SIZE>::decode_with_param(&(false, verifier_len), bytes)?,
+            )),
+            (None, None) => Ok(None),
+            (Some(_), None) => Err(CodecError::UnexpectedValue),
+        }?;
+        Ok(Self {
+            vidpf_proof,
+            szk_query_share_opt,
+        })
+    }
+}
+
+/// Mastic prepare message.
+///
+/// Result of preprocessing the broadcast messages of both aggregators during the
+/// preparation phase.
+pub type MasticPrepareMessage<F, const SEED_SIZE: usize> = Option<SzkVerifier<F, SEED_SIZE>>;
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticPrepareMessage<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        match self {
+            Some(verifier) => verifier.encode(bytes),
+            None => Ok(()),
+        }
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        match self {
+            Some(verifier) => verifier.encoded_len(),
+            None => Some(0),
+        }
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ParameterizedDecode<MasticPrepareState<F, SEED_SIZE>>
+    for MasticPrepareMessage<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        prep_state: &MasticPrepareState<F, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match (&prep_state.szk_query_state, prep_state.verifier_len) {
+            (Some(_), Some(verifier_len)) => Ok(Some(
+                SzkVerifier::<F, SEED_SIZE>::decode_with_param(&(true, verifier_len), bytes)?,
+            )),
+            (None, Some(verifier_len)) => Ok(Some(SzkVerifier::<F, SEED_SIZE>::decode_with_param(
+                &(false, verifier_len),
+                bytes,
+            )?)),
+            (_, None) => Ok(None),
+        }
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Aggregator<SEED_SIZE, NONCE_SIZE> for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    type PrepareState = MasticPrepareState<T::Field, SEED_SIZE>;
+    type PrepareShare = MasticPrepareShare<T::Field, SEED_SIZE>;
+    type PrepareMessage = MasticPrepareMessage<T::Field, SEED_SIZE>;
+
+    fn prepare_init(
+        &self,
+        verify_key: &[u8; SEED_SIZE],
+        agg_id: usize,
+        agg_param: &MasticAggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &MasticPublicShare<VidpfWeight<T::Field>>,
+        input_share: &MasticInputShare<T::Field, SEED_SIZE>,
+    ) -> Result<
+        (
+            MasticPrepareState<T::Field, SEED_SIZE>,
+            MasticPrepareShare<T::Field, SEED_SIZE>,
+        ),
+        VdafError,
+    > {
+        let id = match agg_id {
+            0 => Ok(VidpfServerId::S0),
+            1 => Ok(VidpfServerId::S1),
+            _ => Err(VdafError::Uncategorized(
+                "Invalid aggregator ID".to_string(),
+            )),
+        }?;
+        let mut xof = P::init(
```

Comment:
This includes the onehot proof, but if I'm reading the code correctly, we're missing the path and counter consistency checks: https://jimouris.github.io/draft-mouris-cfrg-mastic/draft-mouris-cfrg-mastic.html#section-3-9

If you prefer we can do this in a future PR.

Reply:
I believe we're leaving this to a future PR